src/pu-main.c

```diff
@@ -5,168 +5,237 @@
-    } else {
-        g_set_error(error, G_OPTION_ERROR, G_OPTION_ERROR_FAILED,
-                    "%s", "Excess values in remaining arguments!");
+    args = pu_command_context_get_args(context);
+    package_path = g_strdup(args[0]);
+    device_path = g_strdup(args[1]);
+
+    if (g_strcmp0(device_path, "") <= 0) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "No device specified");
         return FALSE;
     }
 
+    if (!pu_is_drive(device_path)) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "Device '%s' is not a drive", device_path);
+        return FALSE;
+    }
+
+    if (!pu_device_mounted(device_path, &is_mounted, error)) {
+        g_prefix_error(error, "Failed checking if device is in use: ");
+        return FALSE;
+    }
+
+    if (is_mounted) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "Device '%s' is in use", device_path);
+        return FALSE;
+    }
+
+    if (!pu_package_mount(package_path, &mount_path, error))
+        return FALSE;
+
+    config_path = pu_package_get_layout_file(mount_path, error);
+    config = pu_config_new_from_file(config_path, error);
+    if (config == NULL) {
+        g_prefix_error(error, "Failed creating configuration object for file '%s': ",
+                       config_path);
+        return error_out(mount_path);
+    }
+    api_version = pu_config_get_api_version(config);
+    if (api_version > PARTUP_VERSION_MAJOR) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "API version %d of configuration file is not compatible "
+                    "with program version %d", api_version, PARTUP_VERSION_MAJOR);
+        return error_out(mount_path);
+    }
+
+    emmc = pu_emmc_new(device_path, config, mount_path,
+                       arg_install_skip_checksums, error);
+    if (emmc == NULL) {
+        g_prefix_error(error, "Failed parsing eMMC info from config: ");
+        return error_out(mount_path);
+    }
+    if (!pu_flash_init_device(PU_FLASH(emmc), error)) {
+        g_prefix_error(error, "Failed initializing device: ");
+        return error_out(mount_path);
+    }
+    if (!pu_flash_setup_layout(PU_FLASH(emmc), error)) {
+        g_prefix_error(error, "Failed setting up layout on device: ");
+        return error_out(mount_path);
+    }
+    if (!pu_flash_write_data(PU_FLASH(emmc), error)) {
+        g_prefix_error(error, "Failed writing data to device: ");
+        pu_umount_all(device_path, NULL);
+        return error_out(mount_path);
+    }
+
+    return pu_package_umount(mount_path, error);
+}
+
+static gboolean
+cmd_package(PuCommandContext *context,
+            GError **error)
+{
+    g_autofree gchar *package = NULL;
+    g_autoptr(GPtrArray) input_files = NULL;
+    gchar **args;
+
+    args = pu_command_context_get_args(context);
+    package = g_strdup(args[0]);
+    input_files = g_ptr_array_new();
+    for (gint i = 1; args[i] != NULL; i++)
```

Comment:
Has args a NULL sentinel or do you need to check for length also?

Reply:
Yes, `pu_command_context_get_args()` returns `context->args` which is guaranteed to be a NULL-terminated array of string by the usage of `g_strdupv()`.